futures-util/src/future/shared.rs

```diff
@@ -124,8 +124,14 @@ impl<F> Shared<F> where F: Future {
     }
 
     /// Registers the current task to receive a wakeup when `Inner` is awoken.
-    fn set_waker(&mut self, cx: &mut task::Context) {
+    fn set_waker_if_not_complete(&mut self, cx: &mut task::Context) {
+        // Acquire the lock first before checking COMPLETE to ensure there
+        // isn't a race.
         let mut wakers = self.inner.notifier.wakers.lock().unwrap();
+        if COMPLETE == self.inner.notifier.state.load(SeqCst) {
```

Comment:
we can save this atomic op if we put the wakers into an option :D hehe ^^'

Reply:
lol sure XD

Reply:
Done.